README

				  Net::Zemanta
				  ============


Net::Zemanta is a Perl interface to services provided by Zemanta.

Zemanta is building a platform for assisted on-line content production.
The system analyses the content and returns suggested images, smart links,
keywords and relevant related stories from the Internet.

See http://zemanta.com for more information.

This module currently allows you to use the Suggest service and retrieve user
configuration for the service.


Installation
============

To install this module run the following commands:

   perl Makefile.PL
   make
   make test
   make install


Dependencies
============

  libwww-perl
  JSON


Contact
=======

Please contact the author in case of bugs, patches, etc. at tomaz@zemanta.com


License
=======

Copyright (C) 2008 by Zemanta ltd.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, either Perl version 5.8.7 or, at your option,
any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Net::Zemanta.

0.6  30 March 2014

	- Add missing prerequisites.

0.5  27 October 2009

	- Don't fail unit tests if Test::Pod::Coverage isn't installed.

0.4  22 October 2009

	- Support for new optional zemanta.suggest parameters.

0.3  13 August 2008

	- Support for zemanta.suggest_markup API call.

0.2  8 August 2008

	- Update for the August version of Zemanta API.
	- Support for "freebase" parameter to zemanta.suggest call

0.1  11 July 2008

	- initial release


lib/Net/Zemanta/Method.pm
package Net::Zemanta::Method;

use warnings;
use strict;

use LWP::UserAgent;
use HTTP::Request::Common;
use Encode;
use JSON;

our $VERSION = '0.6';

my $SERVICE_TYPE = "rest";
my $API_VERSION = "0.0";

my $UA_SUFFIX = "Perl-Net-Zemanta/$VERSION";

sub new {
	my $class = shift;
	my %params = @_;

	my $self = \%params;

	$self->{ua} = LWP::UserAgent->new() or return undef;

	$self->{APIKEY} or return undef;
	$self->{METHOD} or return undef;

	# NOTE: trailing space leaves LWP::UserAgent agent string in place
	my $agent = "$UA_SUFFIX ";
	$agent = "$params{USER_AGENT} $agent" if $params{USER_AGENT};
	$self->{ua}->agent($agent);

	$self->{service_url} = "http://api.zemanta.com/services/$SERVICE_TYPE/$API_VERSION";

	bless $self, $class;

	return $self;
}

sub execute {
	my $self = shift;
	my %params = @_;

	my %merged_params = ( 	'method'  => $self->{METHOD},
				'format'  => 'json',
				'api_key' => $self->{APIKEY},
				%params );

	my $response = $self->{ua}->request(POST $self->{service_url},
						\%merged_params );
	
	unless ($response->is_success) {
		$self->{error} = $response->status_line;
		return undef;
	}

	my $result = from_json($response->content);

	if ($result->{status} ne 'ok') {
		$self->{error} = "method returned " . $result->{status};
		return undef;
	}

	$self->{error} = undef;

	return $result;
}

sub error {
	my $self = shift;

	return $self->{error};
}

1;


lib/Net/Zemanta/Preferences.pm
package Net::Zemanta::Preferences;

=encoding utf8

=head1 NAME

Net::Zemanta::Preferences - Perl interface to Zemanta user preferences

=cut

use warnings;
use strict;

use Net::Zemanta::Method;
our @ISA = qw(Net::Zemanta::Method);

=head1 SYNOPSIS

	use Net::Zemanta::Preferences;

	my $zemanta = Net::Zemanta::Preferences->new(
			APIKEY => 'your-API-key' 
		);

	my $preferences = $zemanta->get();

	# URL of the web page for setting preferences
	$preferences->{config_url}

	# User's Amazon affiliate ID
	$preferences->{config_url}

=head1 METHODS

=over 8

=item B<new()>

	Net::Zemanta::Preferences->new(PARAM => ...);

Acceptable parameters:

=over 4

=item  APIKEY

The API key used for authentication with the service.

=item  USER_AGENT

If supplied the value is prepended to this module's identification string 
to become something like:

	your-killer-app/0.042 Perl-Net-Zemanta/0.1 libwww-perl/5.8

Otherwise just Net::Zemanta's user agent string will be sent.

=back

C<new()> returns C<undef> on error.

=cut

sub new {
	my $class 	= shift;
	my %params	= @_;

	$params{METHOD} = "zemanta.preferences";

	my $self = $class->SUPER::new(%params);

	return unless $self;

	bless ($self, $class);
	return $self;
}

=item B<get()>

Returns current settings for the specified API key in form of a hash reference.

See L<http://developer.zemanta.com> for a list of all available settings.

Returns C<undef> on error.

=item B<error()>

If the last call to C<suggest()> returned an error, this function returns a
string containing a short description of the error. Otherwise it returns
C<undef>.

=back

=cut

sub get {
	my $self = shift;

	return $self->execute();
}

=head1 SEE ALSO

=over 4

=item * L<http://zemanta.com>

=item * L<http://developer.zemanta.com>

=back

=head1 AUTHOR

Tomaž Šolc E<lt>tomaz@zemanta.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Zemanta ltd.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, either Perl version 5.8.7 or, at your option,
any later version of Perl 5 you may have available.

=cut


lib/Net/Zemanta/Suggest.pm
package Net::Zemanta::Suggest;

=head1 NAME

Net::Zemanta::Suggest - Perl interface to Zemanta Suggest service

=cut

use warnings;
use strict;

use Net::Zemanta::Method;
our @ISA = qw(Net::Zemanta::Method);

=head1 SYNOPSIS

	use Net::Zemanta::Suggest;

	my $zemanta = Net::Zemanta::Suggest->new(
			APIKEY => 'your-API-key' 
		);

	my $suggestions = $zemanta->suggest(
			"Cozy lummox gives smart squid who asks for job pen."
		);

	# Suggested images
	for $image (@{$suggestions->{images}}) {
		$image->{url_m};
		$image->{description};
	}

	# Related articles
	for $article (@{$suggestions->{articles}}) {
		$article->{url};
		$article->{title};
	}

	# In-text links
	for $link (@{$suggestions->{markup}->{links}}) {
		for $target (@{$link->{target}}) {
			$link->{anchor}, " -> ", $target->{url};
		}
	}

	# Keywords
	for $keyword (@{$suggestions->{keywords}}) {
		$keyword->{name};
	}

=head1 METHODS

=over 8

=item B<new()>

	Net::Zemanta::Suggest->new(PARAM => ...);

Acceptable parameters:

=over 4

=item  APIKEY

The API key used for authentication with the service.

=item  MARKUP_ONLY

If set to true, a faster variant of the API call is made that only returns
the markup element. Default is to provide everything (images, related articles,
markup and tags).

=item  USER_AGENT

If supplied the value is prepended to this module's identification string 
to become something like:

	your-killer-app/0.042 Perl-Net-Zemanta/0.1 libwww-perl/5.8

Otherwise just Net::Zemanta's user agent string will be sent.

=back

C<new()> returns C<undef> on error.

=cut

sub new {
	my $class 	= shift;
	my %params	= @_;

	if( $params{MARKUP_ONLY} ) {
		$params{METHOD} = "zemanta.suggest_markup";
	} else {
		$params{METHOD} = "zemanta.suggest";
	}

	my $self = $class->SUPER::new(%params);

	return unless $self;

	bless ($self, $class);
	return $self;
}

=item B<suggest()>

	$suggestions = $zemanta->suggest( text, PARAM => ... )

Requests suggestions for the given text. Suggestions are returned as a tree
of hash and list references that correspond to the returned JSON data
structure. The most important elements are described bellow. For the full
reference see L<http://developer.zemanta.com>.

Optional parameters:

=over 4

=item  FREEBASE

Set to 1 to include FreeBase (L<http://www.freebase.com>) GUIDs with in-text
markup.

=back

C<suggest()> returns C<undef> on error.

=over 8

=item B<articles>

Related articles. Contains a list of article objects, each having the following
elements:

=over 4

=item B<url>

URL of the article.

=item B<title>

Title of the article.

=item B<published_datetime>

Date when article was published in ISO 8601 format.

=back

=item B<keywords>

Suggested keywords. Contains a list of keyword objects, each having the
following elements:

=over 4

=item B<name>

Keyword name (may contain spaces)

=back

=item B<images>

Related images. Contains a list of image objects, each having the following
elements:

=over 4

=item B<url_l>, B<url_m>, B<url_s>

URLs of a large, medium and small version of the picture respectively.

=item B<url_l_w>, B<url_l_h>, B<url_m_w>, B<url_m_h>, B<url_s_w>, B<url_s_h>

Width and height of large, medium and small version of the picture respectively.

=item B<license>

String containing license terms.

=item B<description>

String containing description

=item B<attribution>

Attribution that must be posted together with the image.

=item B<source_url>

URL of a web page where more information about the image can be found.

=back

=item B<markup>

An object containing the following elements:

=over 4

=item B<text>

HTML formatted input text with added in-text hyperlinks.

=item B<links>

Suggested in-text hyperlinks. A list of link objects, each having 
the following elements:

=over 4

=item B<anchor>

Word or phrase in the original text that should be used as the anchor for the
link.

=item B<target>

List of possible targets for this link. Each target has the following elements:

=over 4

=item B<url>

Destination URL.

=item B<type>

Type of the resource URL is pointing to.

=item B<title>

Title of the resource.

=back

=item B<freebase_guid>

FreeBase GUID (if C<suggest()> was called with FreeBase support enabled)

=back

=back

=item B<rid>

Request ID.

=item B<signature>

HTML signature that should be appended to the text.

=back

=item B<error()>

If the last call to C<suggest()> returned an error, this function returns a
string containing a short description of the error. Otherwise it returns
C<undef>.

=back

=cut

sub suggest {
	my $self = shift;

	my ($text, %options) = @_;

	$options{'FREEBASE'} = 0 unless defined $options{'FREEBASE'};

	my $text_utf8 = Encode::encode("utf8", $text);

	return $self->execute( text     => $text_utf8,
			       freebase => $options{'FREEBASE'} );
}

=head1 SEE ALSO

=over 4

=item * L<http://zemanta.com>

=item * L<http://developer.zemanta.com>

=back

=head1 AUTHOR

Tomaz Solc E<lt>tomaz@zemanta.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Zemanta ltd.

This library is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, either Perl version 5.8.7 or, at your option,
any later version of Perl 5 you may have available.

=cut
